### PR TITLE
Update comment count based on moderator-approved comments

### LINKED ISSUE #9397: Update comment count only after moderator approval
### Please describe the problem (or idea)

> What happened just before the problem occurred? Or what problem could this idea solve?

The comment count on any note is updated before approvals from moderators.

> What did you expect to see that you didn't?

The comment count should only update when approved by mods.

### Please show us where to look

image
https://publiclab.org/notes/cess/03-24-2021/call-for-2021-gsoc-proposals#Proposals

image
https://publiclab.org/notes/barun1024/03-24-2021/gsoc-2021-proposal-geographic-features-refinement

### What's your PublicLab.org username?

barun1024

### PR DESCRIPTION
**Pull request by reginaalyssa**
Updates column comments_count of model Node to increment/decrement only based on approved comments
Updates comment count on create of a new comment
Updates comment count on delete of an existing comment
Updates comment count on /notes page
Updates test based on new response of delete comment

Fixes #9397

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation (no UI updates)
* [x] ask `@publiclab/reviewers` for help, in a comment below

### CODE REVIEW

**Comment by jywarren** [comment]
My apologies on replying slowly to this - i expect something in the changes we've made is causing a test error and will help debug today!

**Comment by jywarren** [comment]
Almost all the errors look like this:

```
ERROR CommentTest#test_question:_image_DRAG_&_DROP_into_EDIT_form_isn't_cross-wired_with_MAIN_form (206.19s)
Minitest::UnexpectedError:         NoMethodError: undefined method `split' for nil:NilClass
            app/models/comment.rb:90:in `next_thread'
            app/models/node.rb:653:in `add_comment'
            test/comment_system_tests/comment_test.rb:720:in `block (2 levels) in <class:CommentTest>'
```

So let's see -- where could we be introducing a nil value?

**Comment by jywarren** [comment]
This is the line that's failing due to a nil value being split: https://github.com/publiclab/plots2/blob/de395b9062221816907f4b36fadc3daeb150a2ac/app/models/comment.rb#L90

**Comment by jywarren** [comment]
That code is being used here:

https://github.com/publiclab/plots2/blob/a313f1033f322e0c4f593cd935f266c724cc265f/app/models/node.rb#L655-L656

So it's trying to run `next_thread` on `comments.last` which we've just checked for not being nil...? Ah, but here there /is/ a `comment.last` - it just has a `nil` value for `comment.last.thread`

How is `comment.thread` assigned? Did we miss assigning it a value, or do we need to ALSO check for `!comments.last.thread.nil?`?

It has no default set: https://github.com/publiclab/plots2/blob/a313f1033f322e0c4f593cd935f266c724cc265f/db/schema.rb.example#L47

Having looked around, i'm not actually sure we even use `thread` anywhere. Maybe it's a leftover from when it was a Drupal website, 10 years ago? I think the easy thing would be to just check if it's nil and assign it "01/" as in the line above. Can you try modifying the code to assign that just in case? I think that should solve it...

**Comment by reginaalyssa** [comment]
Will try it out thank you for the help @jywarren! 😄 Thanks as well @DalilaBA for your response, really appreciate it! 😊

**Comment by reginaalyssa** [comment]
This PR is ready for review, thank you again so much for the help on the issue I encountered with the test @jywarren 🙏

File: app/models/comment.rb [hunk 1]
```diff
@@ -14,6 +11,8 @@ class Comment < ApplicationRecord
 
   validates :comment, presence: true
 
+  scope :published, -> { where(status: 1) }
```
Comment (jywarren) [inline app/models/comment.rb:14]: nice! :-)

File: app/models/comment.rb [hunk 2]
```diff
@@ -2,10 +2,7 @@ class Comment < ApplicationRecord
   include CommentsShared
   extend RawStats
 
-  after_save :update_counter
-  after_destroy :update_counter
-
-  belongs_to :node, foreign_key: 'nid', touch: true
+  belongs_to :node, foreign_key: 'nid', touch: true, counter_cache: true
```
Comment (jywarren) [inline app/models/comment.rb]: cool!

**Review by jywarren — APPROVED** [review]
These changes look great! Thanks so much!!!

**Comment by jywarren** [comment]
Great work here!!! 🎉

**Comment by jywarren** [comment]
We'd love to work with you on another project if you're interested!! Very nicely done work here.

**Pull request merged**